File: backend/ibkr/halt_eta.py
"""LULD / halt reopen ETA clock (pure).

IBKR incoming tick type 49 (ib_async: ``ticker.halted``; Generic tick
required = "-"; do not put 49 in genericTickList -- #178):
  -1 status unavailable (not a halt)
   0 not halted
   1 general halt (news / regulatory) -- NEWS badge, no timer
   2 volatility halt (LULD) -- second-precise 0-5 / 5-10 / >10 clock

``halt_start`` is the first observed ``ticker.halted`` transition, not the
SIP official start, unless Nasdaq Trade Halt RSS supplies one. A quiet
tape is a hint only -- it never creates a chip or countdown. Display only
-- never block Place from this clock.
"""
from __future__ import annotations

from typing import Any

from constants import LULD_CONFIDENT_WINDOW_SEC, LULD_PAUSE_SEC

KIND_LULD = "luld"
KIND_REGULATORY = "regulatory"
KIND_UNKNOWN = "unknown"

BADGE_LULD = "LULD"
BADGE_NEWS = "NEWS"
BADGE_UNK = "UNK"

PHASE_PAUSE = "luld_pause"
PHASE_AUCTION = "luld_auction"
PHASE_EXTENDED = "luld_extended"

LABEL_EXTENDED = "Extended · no ETA"
LABEL_REGULATORY = "NEWS · HALTED"
LABEL_UNKNOWN = "UNK · HALTED"
# Backward-compatible alias -- old tests / imports used LABEL_STILL.
LABEL_STILL = LABEL_EXTENDED

RULE_LULD = (
    "halt_start + 5m LULD pause, then ~5m auction. "
    "After 10m still halted: no confident countdown "
    "(further +5m auction windows possible)."
)
RULE_REGULATORY = "News / regulatory halt -- no timed reopen estimate."
RULE_UNKNOWN = "Halt signaled; type unknown -- no timed reopen estimate."

REASON_LULD = "Volatility pause (ticker.halted=2, incoming tick type 49)"
REASON_REGULATORY = (
    "General halt -- news / regulatory (ticker.halted=1, incoming tick type 49)"
)
REASON_UNKNOWN = "Halt signaled; ticker.halted type unknown"

HALT_START_SOURCE = "observed_ticker_halted"
HALT_START_SOURCE_NASDAQ = "nasdaq_trade_halt_rss"

EXCHANGE_OK = "ok"
EXCHANGE_PENDING = "pending"
EXCHANGE_DOWN = "down"


def parse_halt_code(raw: Any) -> int | None:
    """Return the integer ticker.halted code, or None when missing / NaN."""
    if raw is None:
        return None
    try:
        val = float(raw)
    except (TypeError, ValueError):
        return None
    if val != val:  # NaN
        return None
    return int(val)


def classify_rss_reason_code(reason_code: str | None) -> str:
    """Map a Nasdaq Trade Halt reason to a chip kind (LUDP/T1/…)."""
    raw = (reason_code or "").strip().upper()
    if not raw:
        return KIND_UNKNOWN
    if raw.startswith("LUD") or raw == "LULD" or raw.startswith("VOL"):
        return KIND_LULD
    if raw.startswith("T") or raw.startswith("N"):
        return KIND_REGULATORY
    return KIND_UNKNOWN


def classify_halt_code(code: int | None) -> str | None:
    """Map a ticker.halted code to a chip kind, or None when not halted.

    IBKR Halted values (incoming tick type 49): 2 is LULD, 1 is general/news,
    -1 is unavailable, 0 is not halted. Any other positive code is treated
    as halted-unknown rather than a fake LULD timer.
    """
    if code is None or code in (0, -1):
        return None
    if code == 2:
        return KIND_LULD
    if code == 1:
        return KIND_REGULATORY
    return KIND_UNKNOWN


def kind_badge(kind: str | None) -> str:
    if kind == KIND_LULD:
        return BADGE_LULD
    if kind == KIND_REGULATORY:
        return BADGE_NEWS
    return BADGE_UNK


def format_clock(elapsed_sec: float) -> str:
    """Second-precise elapsed / countdown (M:SS or H:MM:SS)."""
    total = max(0, int(elapsed_sec))
    minutes, seconds = divmod(total, 60)
    if minutes >= 60:
        hours, minutes = divmod(minutes, 60)
        return f"{hours}:{minutes:02d}:{seconds:02d}"
    return f"{minutes}:{seconds:02d}"


def luld_phase(elapsed_sec: float) -> str:
    if elapsed_sec < LULD_PAUSE_SEC:
        return PHASE_PAUSE
    if elapsed_sec < LULD_CONFIDENT_WINDOW_SEC:
        return PHASE_AUCTION
    return PHASE_EXTENDED


def luld_label(
    elapsed_sec: float,
    *,
    start_late: bool = False,
    has_official_start: bool = False,
) -> str:
    clock = format_clock(elapsed_sec)
    confident = (not start_late) or has_official_start
    if not confident:
        return f"{BADGE_LULD} · {clock}"
    phase = luld_phase(elapsed_sec)
    if phase == PHASE_PAUSE:
        left = format_clock(LULD_PAUSE_SEC - elapsed_sec)
        return f"{BADGE_LULD} · {clock} · {left} left"
    if phase == PHASE_AUCTION:
        left = format_clock(LULD_CONFIDENT_WINDOW_SEC - elapsed_sec)
        return f"Auction · {clock} · {left} left"
    return LABEL_EXTENDED


def halt_chip_view(
    *,
    kind: str | None,
    halt_start: float | None,
    now: float,
    halted: bool = True,
    start_late: bool = False,
    official_halt_start: float | None = None,
) -> dict[str, Any] | None:
    """Pure chip view. Returns None when the chip must not render."""
    if not halted or not kind:
        return None

    clock_start = halt_start
    clock_source = HALT_START_SOURCE
    if official_halt_start is not None:
        clock_start = official_halt_start
        clock_source = HALT_START_SOURCE_NASDAQ

    elapsed = 0.0
    if clock_start is not None:
        elapsed = max(0.0, float(now) - float(clock_start))

    has_official = official_halt_start is not None
    badge = kind_badge(kind)

    if kind == KIND_LULD:
        label = luld_label(
            elapsed, start_late=start_late, has_official_start=has_official,
        )
        rule = RULE_LULD
        reason = REASON_LULD
        halt_type = "LULD / volatility pause"
        phase = luld_phase(elapsed)
    elif kind == KIND_REGULATORY:
        label = LABEL_REGULATORY
        rule = RULE_REGULATORY
        reason = REASON_REGULATORY
        halt_type = "News / regulatory"
        phase = KIND_REGULATORY
    else:
        label = LABEL_UNKNOWN
        rule = RULE_UNKNOWN
        reason = REASON_UNKNOWN
        halt_type = "Unknown"
        phase = KIND_UNKNOWN

    return {
        "kind": kind,
        "badge": badge,
        "phase": phase,
        "label": label,
        "elapsed_sec": elapsed,
        "halt_type": halt_type,
        "reason": reason,
        "rule": rule,
        "halt_start_source": clock_source,
        "start_late": bool(start_late),
        "has_official_start": has_official,
    }

File: frontend/src/ibkr/haltEta.ts
/** Pure LULD / halt chip clock -- mirrors backend/ibkr/halt_eta.py. */
import type { HaltSnapshot } from '../types/ticker';
import {
  HALT_BADGE_LULD,
  HALT_BADGE_NEWS,
  HALT_BADGE_UNK,
  HALT_EXCHANGE_PENDING_NOTE,
  HALT_KIND_LULD,
  HALT_KIND_REGULATORY,
  HALT_KIND_UNKNOWN,
  HALT_LABEL_EXTENDED,
  HALT_LABEL_REGULATORY,
  HALT_LABEL_UNKNOWN,
  HALT_RULE_LULD,
  HALT_RULE_REGULATORY,
  HALT_RULE_UNKNOWN,
  HALT_START_LATE_NOTE,
  HALT_START_OBSERVED_NOTE,
  HALT_TYPE_LULD,
  HALT_TYPE_REGULATORY,
  HALT_TYPE_UNKNOWN,
  LULD_CONFIDENT_WINDOW_SEC,
  LULD_PAUSE_SEC,
} from '../constantGroups/halt_eta';
import { STOCK_VIEW_CLOCK_TIMEZONE } from '../constantGroups/chart_api';

export type HaltChipView = {
  kind: string;
  badge: string;
  phase: string;
  label: string;
  tooltip: string;
  startLate: boolean;
  exchangeStatus: string;
};

export function formatClock(elapsedSec: number): string {
  const total = Math.max(0, Math.floor(elapsedSec));
  const minutes = Math.floor(total / 60);
  const seconds = total % 60;
  if (minutes >= 60) {
    const hours = Math.floor(minutes / 60);
    const rem = minutes % 60;
    return `${hours}:${String(rem).padStart(2, '0')}:${String(seconds).padStart(2, '0')}`;
  }
  return `${minutes}:${String(seconds).padStart(2, '0')}`;
}

export function luldPhase(elapsedSec: number): string {
  if (elapsedSec < LULD_PAUSE_SEC) return 'luld_pause';
  if (elapsedSec < LULD_CONFIDENT_WINDOW_SEC) return 'luld_auction';
  return 'luld_extended';
}

export function luldLabel(
  elapsedSec: number,
  startLate = false,
  hasOfficialStart = false,
): string {
  const clock = formatClock(elapsedSec);
  const confident = !startLate || hasOfficialStart;
  if (!confident) return `${HALT_BADGE_LULD} · ${clock}`;
  const phase = luldPhase(elapsedSec);
  if (phase === 'luld_pause') {
    return `${HALT_BADGE_LULD} · ${clock} · ${formatClock(LULD_PAUSE_SEC - elapsedSec)} left`;
  }
  if (phase === 'luld_auction') {
    return `Auction · ${clock} · ${formatClock(LULD_CONFIDENT_WINDOW_SEC - elapsedSec)} left`;
  }
  return HALT_LABEL_EXTENDED;
}

function formatHaltStartEt(haltStart: number | null | undefined, note: string): string {
  if (haltStart == null || !Number.isFinite(haltStart)) {
    return `unknown (${note})`;
  }
  try {
    return `${new Date(haltStart * 1000).toLocaleString('en-US', {
      timeZone: STOCK_VIEW_CLOCK_TIMEZONE,
      month: 'short',
      day: 'numeric',
      year: 'numeric',
      hour: '2-digit',
      minute: '2-digit',
      second: '2-digit',
    })} ET (${note})`;
  } catch {
    return `${haltStart} (${note})`;
  }
}

function formatResumeEt(ts: number | null | undefined): string | null {
  if (ts == null || !Number.isFinite(ts)) return null;
  try {
    return `${new Date(ts * 1000).toLocaleString('en-US', {
      timeZone: STOCK_VIEW_CLOCK_TIMEZONE,
      hour: '2-digit',
      minute: '2-digit',
      second: '2-digit',
    })} ET`;
  } catch {
    return null;
  }
}

function tooltipFor(halt: HaltSnapshot, haltType: string, rule: string): string {
  const lines = [
    `Halt type: ${haltType}`,
    `Halt start: ${formatHaltStartEt(halt.halt_start, HALT_START_OBSERVED_NOTE)}`,
    `Rule: ${rule}`,
  ];
  if (halt.start_late) lines.push(`Honesty: ${HALT_START_LATE_NOTE}`);
  if (halt.reason) lines.push(`Reason: ${halt.reason}`);
  const ex = halt.exchange;
  if (!ex || !ex.matched) {
    lines.push(HALT_EXCHANGE_PENDING_NOTE);
    return lines.join('\n');
  }
  if (ex.reason_code) lines.push(`Nasdaq reason: ${ex.reason_code}`);
  if (ex.pause_threshold) lines.push(`Pause threshold: ${ex.pause_threshold}`);
  if (ex.official_halt_start != null) {
    lines.push(`Nasdaq halt start: ${formatHaltStartEt(ex.official_halt_start, 'Nasdaq Trade Halt RSS')}`);
  }
  const quote = formatResumeEt(ex.quote_resume);
  const trade = formatResumeEt(ex.trade_resume);
  if (quote) lines.push(`Nasdaq quote resume: ${quote}`);
  if (trade) lines.push(`Nasdaq trade resume: ${trade}`);
  if (!quote && !trade) {
    lines.push('Nasdaq scheduled resume: none listed (not invented)');
  }
  return lines.join('\n');
}

export function haltChipView(
  halt: HaltSnapshot | null | undefined,
  nowMs: number,
): HaltChipView | null {
  if (!halt?.halted || !halt.kind) return null;
  const nowSec = nowMs / 1000;
  const official = halt.exchange?.official_halt_start ?? null;
  const hasOfficial = official != null && Number.isFinite(official);
  const clockStart = hasOfficial ? official : halt.halt_start;
  const elapsed = clockStart != null ? Math.max(0, nowSec - clockStart) : 0;
  const startLate = Boolean(halt.start_late);
  const exchangeStatus = halt.exchange?.status ?? 'pending';
  const reason = halt.reason ?? null;
  const ruleFromWire = halt.rule ?? null;

  if (halt.kind === HALT_KIND_LULD) {
    return {
      kind: HALT_KIND_LULD,
      badge: HALT_BADGE_LULD,
      phase: luldPhase(elapsed),
      label: luldLabel(elapsed, startLate, hasOfficial),
      startLate,
      exchangeStatus,
      tooltip: tooltipFor(halt, HALT_TYPE_LULD, ruleFromWire ?? HALT_RULE_LULD),
    };
  }
  if (halt.kind === HALT_KIND_REGULATORY) {
    return {
      kind: HALT_KIND_REGULATORY,
      badge: HALT_BADGE_NEWS,
      phase: HALT_KIND_REGULATORY,
      label: HALT_LABEL_REGULATORY,
      startLate,
      exchangeStatus,
      tooltip: tooltipFor(
        { ...halt, reason },
        HALT_TYPE_REGULATORY,
        ruleFromWire ?? HALT_RULE_REGULATORY,
      ),
    };
  }
  return {
    kind: HALT_KIND_UNKNOWN,
    badge: HALT_BADGE_UNK,
    phase: HALT_KIND_UNKNOWN,
    label: HALT_LABEL_UNKNOWN,
    startLate,
    exchangeStatus,
    tooltip: tooltipFor(halt, HALT_TYPE_UNKNOWN, ruleFromWire ?? HALT_RULE_UNKNOWN),
  };
}
